core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala

```diff
@@ -38,33 +38,41 @@ private[zio] final class FiberContext[E, A](
   private[this] val state = new AtomicReference[FiberState[E, A]](FiberState.Initial[E, A])
 
   // Accessed from within a single thread (not necessarily the same):
-  @volatile private[this] var noInterrupt = 0
   @volatile private[this] var supervised  = List.empty[Set[FiberContext[_, _]]]
   @volatile private[this] var supervising = 0
   @volatile private[this] var locked      = List.empty[Executor]
   @volatile private[this] var environment = List[Any](())
 
-  private[this] val fiberId = FiberContext.fiberCounter.getAndIncrement()
-  private[this] val stack   = new Stack[Any => IO[Any, Any]]()
+  private[this] val interruptStatus = StackBool()
+  private[this] val fiberId         = FiberContext.fiberCounter.getAndIncrement()
+  private[this] val stack           = new Stack[Any => IO[Any, Any]]()
 
   final def runAsync(k: Callback[E, A]): Unit =
     register0(xx => k(Exit.flatten(xx))) match {
       case null =>
       case v    => k(v)
     }
 
+  private object InterruptExit extends Function[Any, IO[E, Any]] {
```

Comment:
Why have `interruptStatus` at all when you can have
```scala
private[this] var interruptible = true
private class InterruptExit(val wasInterruptible: Boolean) extends Function[Any, IO[E, Any]] {
  final def apply(v: Any): IO[E, Any] = {
    interruptible = wasInterruptible
    ZIO.succeed(v)
  }
}
```
and `interruptStatus.peekOrElse(true)` becomes just `interruptible`?

Reply:
Performance reasons. Compare heap allocations: pushing/popping booleans from `interruptStack` will require no heap and is super fast; and pushing the single `object` that is `InterruptExit` onto the pre-allocated stack also involves no allocations.

Reply:
There is no need for allocations:
```scala
private object InterruptExit {
  val WasInterruptible = new InterruptExit(true)
  val WasNotInterruptible = new InterruptExit(false)
  def apply(wasInterruptible: Boolean): InterruptExit = if (wasInterruptible) WasInterruptible else WasNotInteruptible
}
```

Reply:
Do we always statically know what the state should be at the time we add an interrupt change to the stack? I'll look at that. The current approach does not know, it just pops off the last boolean on the stack.

Reply:
Yes, we do. Pops happen in `InterruptExit.apply and in `unwindStack`. In both cases we have the `InterruptExit` instance.